warp/yul/ToCairoVisitor.py

```diff
@@ -1,114 +1,209 @@
+    "starkware.cairo.common.dict_access": {"DictAccess"},
+    "starkware.cairo.common.math_cmp": {"is_le"},
+    "starkware.cairo.common.default_dict": {
+        "default_dict_new",
+    },
+    "starkware.cairo.common.uint256": {"Uint256", "uint256_eq"},
+    "starkware.cairo.common.cairo_builtins": {"HashBuiltin"},
+    "starkware.starknet.common.storage": {"Storage"},
+    "evm.utils": {"update_msize"},
+}
+
+RETURNS_TWO = [
+    "uint256_add",
+    "uint256_mul",
+    "uint256_unsigned_div_rem",
+    "uint256_signed_div_rem",
+]
 
 class ToCairoVisitor(AstVisitor):
     def __init__(self):
         super().__init__()
         self.repr_stack: list[str] = []
+        self.preamble: bool = False
         self.n_names: int = 0
+        self.imports = defaultdict(set)
+        merge_imports(self.imports, COMMON_IMPORTS)
+        self.cairo_code: str = ""
 
     def translate(self, node: ast.Node) -> str:
         main_part = self.print(node)
         return main_part
 
     def print(self, node: ast.Node, *args, **kwargs) -> str:
-        self.visit(node, *args, **kwargs)
-        return self.repr_stack.pop()
+        return self.visit(node, *args, **kwargs)
 
     def common_visit(self, node: ast.Node, *args, **kwargs):
         raise AssertionError(
             f"Each node type should have a custom visit, but {type(node)} doesn't"
         )
 
-    def visit_typed_name(self, node: ast.TypedName):
-        self.repr_stack.append(f"{node.name} : {node.type}")
+    def visit_typed_name(self, node: ast.TypedName) -> str:
+        return f"{node.name} : {node.type}"
 
-    def visit_literal(self, node: ast.Literal):
+    def visit_literal(self, node: ast.Literal) -> str:
         v = int(node.value)  # to convert bools: True -> 1, False -> 0
         high, low = divmod(v, UINT128_BOUND)
-        self.repr_stack.append(f"Uint256(low={low}, high={high})")
-
-    def visit_identifier(self, node: ast.Identifier):
-        self.repr_stack.append(f"{node.name}")
-
-    def visit_assignment(self, node: ast.Assignment):
-        ids_repr = ", ".join("local " + self.print(x) for x in node.variable_names)
-        value_repr = self.print(node.value)
+        return f"Uint256(low={low}, high={high})"
+
+    def visit_identifier(self, node: ast.Identifier) -> str:
+        return f"{node.name}"
+
+    def generate_ids_typed(self, variable_names, function_name: str = ""):
+        variables = []
+        for var in variable_names:
+            var_repr = self.print(var)
+            if "Uint256" not in var_repr:
+                var_repr += ": Uint256"
+            if function_name in RETURNS_TWO:
+                var_repr += ", _"
+            variables.append(var_repr)
+        return ", ".join("local " + x for x in variables)
+
+    def visit_assignment(self, node: ast.Assignment) -> str:
+        self.preamble = False
+        value_repr: str = self.print(node.value)
+        function_name: str = value_repr[: value_repr.find("(")].strip()
```

Comment:
We can get `function_name` without text manipulation. `function_name = node.value.function_name.name`. It should be done in the if-branch where we make sure that `isinstance(node.value, ast.FunctionCall)`.